### DISCUSSION TITLE
Rules for date computation

### DISCUSSION THREAD

**Discussion by priyankmav** [zeroSteiner/rule-engine#18] [Ideas]
Hello zeroSteiner,

I also have a suggestion for date related rules.
Eg. the rule engine should be able to compute date difference ->
rule string like "($today - effectiveTMS.date).days < 90"

```
{
	"personFacts": [{
		"abc": false,
		"bcd": false,
		"edg": false,
		"asd": false,
		"effectiveTMS": {
			"$date": "1970-01-01T05:00:00.000Z"
		},
		"jjf": 0,
		"sha": false,
		"ncb": false,
		"qwe": false,
		"ytu": 0,
		"aops": 0
	}]
}
```

Let me know if this suggestion helps!

Thanks!

**Comment by priyankmav** [comment]
Hi Spencer, 

@zeroSteiner 
Could you kindly help on this one? If I could subtract two dates and check if its greater or lesser than X days.

Thanks
Priyank Mavani

**Reply by zeroSteiner** [reply]
Would an attribute converting the date into the epoch work for you? You could then subtract the two and convert the desired delta into seconds. Thinking something like this for example:

`(date1.epoch - date2.epoch) > (3 * EPOCH_DAYS)`

**Reply by priyankmav** [reply]
Just to get a little more understanding here, with an example, where i want to check the difference between 2 dates is greater than 90 days would look something like this below (considering 1 EPOCH day = 86400 seconds) -->
`(date1.epoch - date2.epoch) > 90 * 86400` ?

**Reply by zeroSteiner** [reply]
Yeah that's the idea. It would save me from having to add a whole new time delta data type which would be a lot more work.

**Reply by priyankmav** [reply]
I understand. Your proposed solution looks good. Is this already a part of rule-engine, or you would intend to add this in the next release?

**Reply by zeroSteiner** [reply]
It's not, I'll need to add it but I can probably get it done within the next week or two (no promises though).

FWIW I'd also suggest if you're going to do this a lot, adding a variable or even [adding a builtin](https://zerosteiner.github.io/rule-engine/getting_started.html#changing-builtin-symbols) variable that contains the number of seconds in a week, day, hour, minute, etc.

That way you could write your rules and have them be a bit more readable similar to my original example (`(date1.epoch - date2.epoch) > (3 * $days`) where `$days` is an additional builtin variable as noted by the `$` prefix that equals `60 * 60 * 24`. Totally optional though, it's up to you. Using a new builtin variable will ensure that the object your're evaluating doesn't overwrite the `days` symbol because it's in a different scope and you also wouldn't need to define it for all of your objects.

**Reply by priyankmav** [reply]
Yes, I see how adding an additional builtin variable can make the rule syntax cleaner and readable. Again, thank you for your quick response and being so prompt here. I hope these suggestions also make this package more robust and more feature rich. I look forward to the new changes! Cheers! :)

**Reply by zeroSteiner** [reply]
Just published version [3.3.0](https://pypi.org/project/rule-engine/3.3.0/) which adds the `to_epoch` attribute to the `DATETIME` data type.

**Reply by priyankmav** [reply]
That was quick! Thank you! :)
